src/sinks/nats.rs

```diff
@@ -0,0 +1,152 @@
+use crate::{
+    config::{DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
+    emit,
+    event::Event,
+    internal_events::NatsEventSent,
+    sinks::util::StreamSink,
+};
+use async_trait::async_trait;
+use futures::{future, stream::BoxStream, FutureExt, StreamExt};
+use serde::{Deserialize, Serialize};
+
+/**
+ * Code dealing with the SinkConfig struct.
+ *
+ * DEV: Start with the bare minimum for now.
+ */
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct NatsSinkConfig {
+    url: String,
+    subject: String,
+}
+
+inventory::submit! {
+    SinkDescription::new::<NatsSinkConfig>("nats")
+}
+
+impl GenerateConfig for NatsSinkConfig {}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "nats")]
+impl SinkConfig for NatsSinkConfig {
+    async fn build(
+        &self,
+        _cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let sink = NatsSink::new(self.clone());
+        let healthcheck = future::ok(()).boxed();
+
+        Ok((super::VectorSink::Stream(Box::new(sink)), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "nats"
+    }
+}
+
+/**
+ * Code dealing with the Sink struct.
+ */
+
+#[derive(Clone)]
+pub struct NatsSink {
+    nc: nats::Connection,
+    subject: String,
+}
+
+impl NatsSink {
+    fn new(config: NatsSinkConfig) -> Self {
+        let options = nats::Options::new();
+        let nc = options.connect(&config.url).unwrap();
+
+        Self {
+            nc,
+            subject: config.subject,
+        }
+    }
+}
+
+#[async_trait]
+impl StreamSink for NatsSink {
+    async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
+        while let Some(event) = input.next().await {
+            match event {
+                Event::Log(log) => {
+                    let body = log
+                        .get(crate::config::log_schema().message_key())
+                        .map(|v| v.to_string_lossy())
+                        .unwrap_or_else(|| "".into());
+
+                    let message_len = body.len();
+
+                    if self.nc.publish(&self.subject, body).is_ok() {
```

Comment:
I'll note this in the review description, but we should avoid dropping messages that cannot be published. Instead we should continuously retry them, acking (via the sink context) once they were published. This'll give users higher delivery guarantees.

Reply:
I think this is done. I set up the acker and acked on successful processing.
I didn't do anything in the ways of managing a message buffer myself. Is that something the acker handles?

Reply:
:+1: the acking looks good. The message buffer should be handled automatically with that.